enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,421 @@
+clusters so API and ingress are accessible from either ip version.
+
+## Motivation
+
+Customer applications on dual stack clusters may only have ipv6 connectivity,
+but the cluster VIPs will be ipv4 (at this time ipv6 VIPs are not supported
+in dual stack clusters).
+
+### Goals
+
+Create both ipv4 and ipv6 VIPs on dual stack clusters so single stack
+applications have the connectivity they require.
+
+### Non-Goals
+
+* Making the `kubernetes.default` Service dual stack is not supported and will
+not be addressed by this work.
+
+* Configuration of any VIPs beyond a second ipv6 one for dual stack clusters.
+MetalLB is a better solution to creating arbitrary loadbalancers.
+
+## Proposal
+
+### User Stories
+
+As a deployer of a dual stack OpenShift cluster, I want to run ipv6-only
+applications that need access to the API and ingress VIPs.
+
+### API Extensions
+
+New fields will have to be added to the
+[platform status section](https://github.com/openshift/api/blob/24043075985b3dc87190174241f3c4d0416b8ea2/config/v1/types_infrastructure.go#L465)
+of the infrastructure object to persist the specified additional VIPs.
+
+### Implementation Details/Notes/Constraints [optional]
+
+This should be a relatively simple feature to implement since most of the
+new code, manifests, and templates will be slight modifications of the
+existing loadbalancer ones. The main thing we need to determine is what the
+structure of the install-config and api changes will be. That will have an
+impact on the implementation of these new VIPs, but any of the designs
+proposed below should be doable. It's mostly a question of picking the one
+that will provide the best user experience and, secondarily, the simplest
+implementation.
+
+### Risks and Mitigations
+
+Minimal risk. This is just adding another VIP, something we already have in
+our deployments. The main concern would be logic errors arising out of
+the need to handle multiple VIPs, which (if they happen) will need to be
+addressed as bugs.
+
+## Design Details
+
+This feature will require changes in a few different components:
+
+Installer - New fields will need to be added to install-config for deployers
+to specify the additional VIPs.
+
+API - New fields will need to be added to the platform status section of the
+infrastructure object.
+
+machine-config-operator - The on-prem loadbalancer templates and manifests
+will need to be updated to support the additional VIPs.
+
+baremetal-runtimecfg - There is logic that depends on the ip version of the
+VIP. If there are multiple VIPs, this will have to be modified to account for
+that. In addition, the render code will need the new values wired in.
+
+assisted-service - the hive API extension is used for agent-based cluster creation as a stand-alone tool ("Infrastructure Operator") and as part of RHACM. It will need new API fields that correspond to the changes in the install-config.
```

Comment:
Given that we will need to have backward compatibility code in the installer and MCO to handle existing install-configs and clusters, is this a blocker?

I ask because this is already at high risk of missing 4.11 and we're not likely to be able to absorb a significant amount of additional work. If we could land the underlying feature in 4.11 and wire it in to assisted in 4.12 that would help a lot.

Reply:
No no, this is not a blocker.

Also, all the work I described above will be handled by the Assisted Installer team, so on your side no additional work would be needed. I just wanted to make sure that changes required in Assisted are written down somewhere at the moment of accepting this change of the structure.

Assisted Installer releases with ACM and not with OCP, so our freezes are usually around a month later than OCPs. So if you deliver for 4.11, Assisted will catch-up around ACM 2.6

Reply:
Excellent, thank you! I will get this added.